### PR TITLE
Widget set/up/free ops for IPC4

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/sof/ipc4-topology.c [hunk 1]
```diff
@@ -1076,6 +1076,129 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
...
+	case snd_soc_dapm_aif_out:
+	{
+		struct sof_ipc4_copier *ipc4_copier = swidget->private;
+
+		ipc_size = ipc4_copier->ipc_config_size;
+		ipc_data = ipc4_copier->ipc_config_data;
+
+		msg = &ipc4_copier->msg;
+		break;
+	}
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1111]: why do we need those braces? They make the code less readable IMHO

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -1076,6 +1076,129 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
...
+		ret = sof_ipc_tx_message(sdev->ipc, &msg, 0, NULL, 0);
+		if (ret < 0)
+			dev_err(sdev->dev, "failed to free pipeline widget %s",
+				swidget->widget->name);
+
+		pipeline->mem_usage = 0;
+		pipeline->state = SOF_IPC4_PIPE_UNINITIALIZED;
+	}
+
+	if (!ret)
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: can we be consistent on if (!ret) or if (ret < 0). pick one.

**Review by plbossart — COMMENTED** [review]
style nit-picks

File: sound/soc/sof/ipc4-topology.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1111]: I cannot declare the variable without those braces within the case

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: removed this now

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/ipc4-topology.c [hunk 3]
```diff
@@ -1076,6 +1076,125 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
...
+		return -EINVAL;
+	}
+
+	if (swidget->id != snd_soc_dapm_scheduler) {
+		pipeline = pipe_widget->private;
+		msg->primary &= ~SOF_IPC4_MOD_INSTANCE_MASK;
+		msg->primary |= SOF_IPC4_MOD_INSTANCE(swidget->instance_id);
+
+		msg->extension |= ipc_size >> 2;
+		msg->extension |= SOF_IPC4_MOD_EXT_DOMAIN(pipeline->lp_mode);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c:1158]: Should these fields need to be cleared first?

File: sound/soc/sof/ipc4-topology.c [hunk 4]
```diff
@@ -1076,6 +1076,125 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
...
+		header = SOF_IPC4_GLB_PIPE_INSTANCE_ID(swidget->pipeline_id);
+		header |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_DELETE_PIPELINE);
+		header |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+		header |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+
+		msg.primary = header;
+
+		ret = sof_ipc_tx_message(sdev->ipc, &msg, 0, NULL, 0);
+		if (ret < 0)
+			dev_err(sdev->dev, "failed to free pipeline widget %s",
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: Missing `\n`?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@ranj063, Only two comments, otherwise looks neat.

**Comment by ranj063** [comment]
@ujfalusi cleared the bit fields for ipc_size and ext_domain now and fixed the missing \n

**Review by plbossart — APPROVED** [review]

**Comment by ranj063** [comment]
SOFCI TEST

**Review by bardliao — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]

**Pull request merged**